src/classes/CW_AudienceCSVProcessHelper.cls

```diff
@@ -0,0 +1,406 @@
+         grp.Description__c = grpName;
+         grpList.Add(grp);
+      }
+      Insert grpList;
+      map<string,Id> reslt= new map<string,Id>();
+      for(CW_Audience_Group__c rec:grpList ){
+      	reslt.put(rec.Name.ToUpperCase(),rec.id);
+      }
+		return reslt;
+   } 
+   
+   
+   //End OF step 2 - Step 3
+   
+   //Create Campiang Rewards - WORKING OK
+   //Add Error chcking on Insert Records
+   private void loadCreateCampaignReward(){
+   	campRewardMap= new Map<ID,CW_Campaign_Reward__c>();
+   	Map<String,CW_Campaign_Reward__c> campFreeBetdMap= new Map<String,CW_Campaign_Reward__c>();
+   	for(CW_CSV_Stage_AudienceMembers__c rec:CsvStageList ){
+   		campFreeBetdMap.Put(rec.FreeBetId__c.ToUpperCase(),null);
+   	}
+   	for(CW_Campaign_Reward__c rec: [SELECT Id,Reward_Library__c, Reward_Library__r.FreeBet_Id__c,Reward_Library__r.Amount__c,Reward_Library__r.Currency__c FROM CW_Campaign_Reward__c
+   												 WHERE Reward_Library__r.FreeBet_Id__c=: campFreeBetdMap.keySet()]){ //Campaign__Offer__r.Campaign__c=:CampaignID AND 
+   		campRewardMap.put(rec.Reward_Library__c,rec);								 	
+   		campFreeBetdMap.put(rec.Reward_Library__r.FreeBet_Id__c.ToUpperCase(),rec);
+   												 	
+   	}
+   	List<CW_Campaign_Reward__c> CampRewList= new List<CW_Campaign_Reward__c>();
+   	//For percent we takie cuurency from file
+   	
+   	
+   	for(CW_CSV_Stage_AudienceMembers__c rec:CsvStageList ){
+   		String FreeBet=rec.FreeBetId__c.ToUpperCase();
+   		CW_Campaign_Reward__c rewRec= campFreeBetdMap.get(FreeBet);
+   		if(rewRec==null){
+	   		//rewRec=new CW_Campaign_Reward__c(Campaign_Offer__r.Campaign__c = CampaignID);
+	   		rewRec.Reward_Library__c=rec.Reward_Library__c;
+	   		//rewRec.Reward_Catalogue__c=rec.Reward_Library__r.Reward_Catalogue__c;
+	   		//rewRec.Reward_Curency__c=rec.Reward_Library__r.Currency__c;
+	   		CampRewList.Add(rewRec);
+	   		campFreeBetdMap.put(FreeBet,rewRec);
+   		}
+   	}
+   	insert CampRewList;
+   	for(CW_Campaign_Reward__c rec : CampRewList){
+   		campRewardMap.put(rec.Reward_Library__c,rec);
+   	}
+   }
+   
+   
+   private void  createLoadAudienceGroupSchdule(){
+   	
+   	grupSheduleMap = new Map<ID,CW_Audience_Group_Schedule__c>();//Reference by group ID.
+   	Map<Id,CW_Audience_Group__c> grupMap= new Map<Id,CW_Audience_Group__c>();
+   	CW_Audience_Group_Schedule__c c = new  CW_Audience_Group_Schedule__c();
+   	for(CW_CSV_Stage_AudienceMembers__c rec:CsvStageList ){
+   		if(!grupSheduleMap.containskey(rec.Audience_Group_ref__c)){
+   			grupSheduleMap.Put(rec.Audience_Group_ref__c,null);
+   			grupMap.put(rec.Audience_Group_ref__c, new CW_Audience_Group__c(id=rec.Audience_Group_ref__c,
+   											is_control_group__c=rec.Audience_Group_ref__r.Is_control_group__c,
+   											control_group__c=rec.Audience_Group_ref__r.control_group__c));
+   		}
+   	}
+   	for (CW_Audience_Group_Schedule__c rec: [SELECT ID,Audience_Group__c FROM CW_Audience_Group_Schedule__c
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.